### PR TITLE
Add support for ALL the operators

### PR DESCRIPTION
**Pull request by christopher-dG**
rip coverage :thinking:

### CODE REVIEW

File: src/sugar.jl [hunk 1]
```diff
@@ -77,12 +81,15 @@ function atset_impl(ex::Expr)
...
-    else
+    elseif haskey(UPDATE_OPERATOR_TABLE, ex.head)
         op = UPDATE_OPERATOR_TABLE[ex.head]
         f = :(_UpdateOp($op,$val))
         quote
             modify($f, $obj, $lens)
         end
+    else
+        msg = "Unsupported operation $(ex.head)"
+        throw(ArgumentError(msg))
```
[the remark this answers is not in the document]
Comment (christopher-dG) [reply src/sugar.jl]: I tried testing with:
```julia
t = T(1, 2)
@test_throws ArgumentError @set t.a
```
But it throws the error before it reaches the test macro it seems.

**Comment by jw3126** [comment]
Another approach would be to get rid of the table and use something like this instead:
```julia
julia> function get_update_operation(sym::Symbol)
           s = String(sym)
           @assert endswith(s, "=")
           Symbol(s[1:end-1])
       end
get_update_operation (generic function with 1 method)

julia> get_update_operation(:(>>=))
:>>
```
What do you think @christopher-dG ?

**Comment by christopher-dG** [comment]
I like that a lot! It also led me to try to break it and I found operators `%=` and `^=` in the process, so not having to maintain a table manually is definitely a great idea :sweat_smile: Should I close this then perhaps?

**Comment by jw3126** [comment]
Nice, I was not aware of these either. Instead of closing would you like to provide the non table variant :heart:?

**Comment by jw3126** [comment]
Thanks for doing this! If no other concerns come up, I will merge on monday.

**Comment by tkf** [comment]
I guess the full list of operators we need is this? https://github.com/JuliaLang/julia/blob/96ce5ba22122280b67991849acdb17b36603224f/src/julia-parser.scm#L9-L11

```lisp
;; note: there are some strange-looking things in here because
;; the way the lexer works, every prefix of an operator must also
;; be an operator.
(define prec-assignment
  (append! (add-dots '(= += -= *= /= //= |\\=| ^= ÷= %= <<= >>= >>>= |\|=| &= ⊻= ≔ ⩴ ≕))
           '(:= ~ $=)))
```

**Comment by tkf** [comment]
This is better https://github.com/JuliaLang/julia/blob/96ce5ba22122280b67991849acdb17b36603224f/src/julia-parser.scm#L113-L116

```lisp
; operators that are special forms, not function names
(define syntactic-operators
  (append! (add-dots '(= += -= *= /= //= |\\=| ^= ÷= %= <<= >>= >>>= |\|=| &= ⊻=))
           '(:= --> $= && |\|\|| |.| ... ->)))
```

**Comment by tkf** [comment]
If we want to avoid having a table, how about something like this?

```julia
function get_update_op(sym::Symbol)
    s = String(sym)
    if !endswith(s, '=') || isdefined(Base, sym)
        # In Julia, update syntax like `x += y` is lowered to `x = x + y`
        # before compilation.  It means that `+=` does not exist as a
        # function.  If `sym` exists in `Base`, it is likely that this
        # is not an assignment operation.
        error("Operation $sym does not look like an assignment.")
    end
    Symbol(s[1:end-1])
end
```

File: src/sugar.jl [hunk 2]
```diff
@@ -53,12 +53,11 @@ function parse_obj_lens(ex)
...
-const UPDATE_OPERATOR_TABLE = Dict(
-:(+=) => +,
-:(-=) => -,
-:(*=) => *,
-:(/=) => /,
-)
+function get_update_op(sym::Symbol)
+    s = String(sym)
+    @assert endswith(s, '=')
+    Symbol(s[1:end-1])
```
[the remark this answers is not in the document]
Comment (jw3126) [reply src/sugar.jl]: Yes, that is an unwanted glitch. But I think it is not an issue in practice?

**Comment by jw3126** [comment]
I am still leaning towards the function (with @tkf improvement) version instead of a table, but I dont't have a strong opinion. What do you guys prefer?

**Comment by tkf** [comment]
The function-based approach sounds good to me. If Julia introduces a new assignment syntax, we get an update automatically (though this is a very hypothetical scenario).

**Comment by jw3126** [comment]
Merged! Thanks @christopher-dG for this improvement!

**Comment by christopher-dG** [comment]
My pleasure! Could we make a release? I have some code that would like to use `&=` and `|=`.

**Comment by jw3126** [comment]
Sure!

**Pull request merged**